kubetest/aks.go

```diff
@@ -0,0 +1,297 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2019-10-01/containerservice"
+	"github.com/Azure/go-autorest/autorest/azure"
+	uuid "github.com/satori/go.uuid"
+	"golang.org/x/crypto/ssh"
+)
+
+type aksDeployer struct {
+	azureCreds    *Creds
+	azureClient   *AzureClient
+	templateUrl   string
+	outputDir     string
+	resourceGroup string
+	resourceName  string
+}
+
+func newAksDeployer() (*aksDeployer, error) {
+	if err := validateAksFlags(); err != nil {
+		return nil, err
+	}
+
+	creds, err := getAzCredentials()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get azure credentials: %v", err)
+	}
+	env, err := azure.EnvironmentFromName(*aksAzureEnv)
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine azure environment: %v", err)
+	}
+
+	var client *AzureClient
+	if client, err = getAzureClient(env,
+		creds.SubscriptionID,
+		creds.ClientID,
+		creds.TenantID,
+		creds.ClientSecret,
+	); err != nil {
+		return nil, fmt.Errorf("error trying to get Azure Client: %v", err)
+	}
+
+	tempdir, err := ioutil.TempDir(os.Getenv("HOME"), "aks")
+	if err != nil {
+		return nil, fmt.Errorf("error creating tempdir: %v", err)
+	}
+
+	return &aksDeployer{
+		azureCreds:    creds,
+		azureClient:   client,
+		templateUrl:   *aksTemplateURL,
+		outputDir:     tempdir,
+		resourceGroup: *aksResourceGroupName,
+		resourceName:  *aksResourceName,
+	}, nil
+}
+
+func validateAksFlags() error {
+	if *aksCredentialsFile == "" {
+		return fmt.Errorf("no credentials file path specified")
+	}
+	if *aksResourceName == "" {
+		*aksResourceName = "kubetest-" + uuid.NewV1().String()
+	}
+	if *aksResourceGroupName == "" {
+		*aksResourceGroupName = *aksResourceName
+	}
+	if *aksDNSPrefix == "" {
+		*aksDNSPrefix = *aksResourceName
+	}
+	return nil
+}
+
+func (a *aksDeployer) Up() error {
+	log.Printf("Creating AKS cluster %v in resource group %v", a.resourceName, a.resourceGroup)
+	templateFile, err := downloadFromURL(a.templateUrl, path.Join(a.outputDir, "kubernetes.json"), 2)
+	if err != nil {
+		return fmt.Errorf("error downloading AKS cluster template: %v with error %v", a.templateUrl, err)
+	}
+
+	template, err := ioutil.ReadFile(templateFile)
+	if err != nil {
+		return fmt.Errorf("failed to read downloaded cluster template file: %v", err)
+	}
+
+	var model containerservice.ManagedCluster
```

Comment:
i'd like to convert this to raw JSON and use a REST client so we can use any version, but for now I will do it with a specific type.